File: services/aas-designer/AasDesignerCommon/Utils/IdGenerationUtil.cs
namespace AasDesignerCommon.Utils;

public enum IdType
{
    Aas,
    Submodel,
    Property,
    Collection,
    Mlp,
    File,
    Asset,
    Other,
    ConceptDescription,
}

public static class IdGenerationUtil
{
    public static string GenerateId(IdType idType, string prefix)
    {
        string generatedIri = prefix + "/ids";
        switch (idType)
        {
            case IdType.Aas:
                generatedIri +=
                    $"/aas/{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}";
                break;
            case IdType.Submodel:
                generatedIri +=
                    $"/sm/{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}";
                break;
            case IdType.Collection:
                generatedIri +=
                    $"/coll/{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}";
                break;
            case IdType.Property:
                generatedIri +=
                    $"/prop/{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}";
                break;
            case IdType.Mlp:
                generatedIri +=
                    $"/mlp/{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}";
                break;
            case IdType.File:
                generatedIri +=
                    $"/file/{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}";
                break;
            case IdType.Asset:
                generatedIri +=
                    $"/asset/{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}";
                break;
            case IdType.Other:
                generatedIri +=
                    $"/other/{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}";
                break;
            case IdType.ConceptDescription:
                generatedIri +=
                    $"/cd/{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}_{GetRandomInt(1000, 9999)}";
                break;
            default:
                throw new ArgumentOutOfRangeException(nameof(idType), idType, null);
        }
        return generatedIri;
    }

    private static int GetRandomInt(int min, int max)
    {
        Random random = new Random();
        return random.Next(min, max + 1);
    }
}

File: frontend/libs/helpers/src/id-generation-util.ts
import { Injectable } from '@angular/core';

@Injectable({
  providedIn: 'root',
})
export class IdGenerationUtil {
  generateIri(
    type: 'submodel' | 'property' | 'collection' | 'mlp' | 'file' | 'asset' | 'other' | string,
    prefix: string,
  ) {
    return IdGenerationUtil.generateIri(type, prefix);
  }

  static generateIri(
    type: 'submodel' | 'property' | 'collection' | 'mlp' | 'file' | 'asset' | 'other' | 'aas' | string,
    prefix: string,
  ) {
    let generatedIRdi = prefix.replace(/\/+$/, '') + '/ids';
    switch (type) {
      case 'submodel':
        generatedIRdi += `/sm/${this.getRandomInt(1000, 9999)}_${this.getRandomInt(1000, 9999)}_${this.getRandomInt(
          1000,
          9999,
        )}_${this.getRandomInt(1000, 9999)}`;
        break;
      case 'aas':
        generatedIRdi += `/aas/${this.getRandomInt(1000, 9999)}_${this.getRandomInt(1000, 9999)}_${this.getRandomInt(
          1000,
          9999,
        )}_${this.getRandomInt(1000, 9999)}`;
        break;
      case 'collection':
        generatedIRdi += `/coll/${this.getRandomInt(1000, 9999)}_${this.getRandomInt(1000, 9999)}_${this.getRandomInt(
          1000,
          9999,
        )}_${this.getRandomInt(1000, 9999)}`;
        break;
      case 'property':
        generatedIRdi += `/prop/${this.getRandomInt(1000, 9999)}_${this.getRandomInt(1000, 9999)}_${this.getRandomInt(
          1000,
          9999,
        )}_${this.getRandomInt(1000, 9999)}`;
        break;
      case 'mlp':
        generatedIRdi += `/mlp/${this.getRandomInt(1000, 9999)}_${this.getRandomInt(1000, 9999)}_${this.getRandomInt(
          1000,
          9999,
        )}_${this.getRandomInt(1000, 9999)}`;
        break;
      case 'file':
        generatedIRdi += `/file/${this.getRandomInt(1000, 9999)}_${this.getRandomInt(1000, 9999)}_${this.getRandomInt(
          1000,
          9999,
        )}_${this.getRandomInt(1000, 9999)}`;
        break;
      case 'asset':
        generatedIRdi += `/asset/${this.getRandomInt(1000, 9999)}_${this.getRandomInt(1000, 9999)}_${this.getRandomInt(
          1000,
          9999,
        )}_${this.getRandomInt(1000, 9999)}`;
        break;
      case 'other':
        generatedIRdi += `/other/${this.getRandomInt(1000, 9999)}_${this.getRandomInt(1000, 9999)}_${this.getRandomInt(
          1000,
          9999,
        )}_${this.getRandomInt(1000, 9999)}`;
        break;
      default:
        generatedIRdi += `/${type}/${this.getRandomInt(1000, 9999)}_${this.getRandomInt(
          1000,
          9999,
        )}_${this.getRandomInt(1000, 9999)}_${this.getRandomInt(1000, 9999)}`;
    }
    return generatedIRdi;
  }

  static getRandomInt(min: number, max: number): number {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }
}
